rpc.go

```diff
@@ -312,3 +310,153 @@ func getMiddlewareResult(u *User, line string) string {
 	}
 	return line
 }
+
+func pluginsCMD(_ string, u *User) {
+	plugins := make([]CMD, 0, len(PluginCMDs))
+	for n, c := range PluginCMDs {
+		plugins = append(plugins, CMD{
+			name:     n,
+			info:     c.info,
+			argsInfo: c.argsInfo,
+		})
+	}
+	autogenerated := autogenCommands(plugins)
+	if autogenerated == "" {
+		autogenerated = "   (no plugin commands are loaded)"
+	}
+	u.room.broadcast("", "Plugin commands  \n"+autogenerated)
+}
+
+type tokensDbEntry struct {
+	Token string `json:"token"`
+	Data  string `json:"data"`
+}
+
+var Tokens = make([]tokensDbEntry, 0)
+
+func initTokens() {
+	if Integrations.RPC == nil {
+		return
+	}
+
+	f, err := os.Open(Config.DataDir + string(os.PathSeparator) + "tokens.json")
+	if err != nil {
+		if !os.IsNotExist(err) {
+			Log.Fatal("Error reading tokens file: " + err.Error())
+		}
+		return
+	}
+	defer f.Close()
+	err = json.NewDecoder(f).Decode(&Tokens)
+	if err != nil {
+		error := "Error decoding tokens file: " + err.Error()
+		MainRoom.broadcast(Devbot, error)
+		Log.Fatal(error)
+	}
+}
+
+func saveTokens() {
+	f, err := os.Create(Config.DataDir + string(os.PathSeparator) + "tokens.json")
+	if err != nil {
+		Log.Fatal("Error saving tokens file: " + err.Error())
+	}
+	defer f.Close()
+	data, err := json.Marshal(Tokens)
+	if err != nil {
+		error := "Error encoding tokens file: " + err.Error()
+		MainRoom.broadcast(Devbot, error)
+		Log.Fatal(error)
+	}
+	_, err = f.Write(data)
+	if err != nil {
+		error := "Error writing tokens file: " + err.Error()
+		MainRoom.broadcast(Devbot, error)
+		Log.Fatal(error)
+	}
+}
+
+func checkToken(token string) bool {
+	if token == Integrations.RPC.Key {
+		return true
+	}
+	for _, t := range Tokens {
+		if t.Token == token {
+			return true
+		}
+	}
+	return false
+}
+
+func lsTokensCMD(_ string, u *User) {
+	if !auth(u) {
+		u.room.broadcast(Devbot, "Not authorized")
+		return
+	}
+
+	if len(Tokens) == 0 {
+		u.writeln(Devbot, "No tokens found.")
+		return
+	}
+	u.writeln(Devbot, "Tokens:")
+	for _, t := range Tokens {
+		u.writeln(Devbot, shasum(t.Token)+"    "+t.Data)
+	}
+}
+
+func revokeTokenCMD(rest string, u *User) {
+	if !auth(u) {
+		u.room.broadcast(Devbot, "Not authorized")
+		return
+	}
+
+	if len(rest) == 0 {
+		u.writeln(Devbot, "Please provide a sha256 hash of a token to revoke.")
+		return
+	}
+	for i, t := range Tokens {
+		if shasum(t.Token) == rest {
+			Tokens = append(Tokens[:i], Tokens[i+1:]...)
+			saveTokens()
+			u.writeln(Devbot, "Token revoked!")
+			return
+		}
+	}
+	u.writeln(Devbot, "Token not found.")
+}
+
+func grantTokenCMD(rest string, u *User) {
+	if !auth(u) {
+		u.room.broadcast(Devbot, "Not authorized")
+		return
+	}
+
+	toUser, ok := findUserByName(u.room, rest)
+	if !ok {
+		// fallback to sending the token to the admin running the cmd
```

Comment:
I would err out at this point and let the admin choose whether to run again with a corrected user arg or without one.

Reply:
I designed it to have this behavior so that you can create separate tokens that aren't associated with users (like if you wanted to `grant-token Hack Club Emoji Middleware`), it does have the cost of the possibility of typos not sending the message to the user but you can always just dm it or revoke/reissue the token

Reply:
You can detect if the admin wants to specify a user by requiring them to use `@user` like some commands do. Then you can respond to typos and also allow for the `grant-token Hack Club Emoji Middleware` use case.